base/task.jl

```diff
@@ -479,14 +485,47 @@ macro async(expr)
         let $(letargs...)
             local task = Task($thunk)
             if $(Expr(:islocal, var))
-                put!($var, task)
+                put!($var, $(wrap(:task)))
             end
             schedule(task)
             task
         end
     end
 end
 
+# task wrapper that doesn't exceptions wrapped in TaskFailedException
+struct UnwrapTaskFailedException
+    task::Task
+end
+
+# common code for wait&fetch for UnwrapTaskFailedException
+macro unwrap_task_failed(f, t)
```

Comment:
Can't this be a function?

Reply:
yeah, will do (I'm not sure what are the overheads of passing functions around to functions, and this pretty much flattened the whole thing and removed the problem :D )